### PR TITLE
Add new endpoint and handler for synchronous prediction requests

### PR DESCRIPTION
**Pull request by acrellin**

### CODE REVIEW

File: cesium_app/handlers/prediction.py [hunk 1]
```diff
@@ -5,11 +5,13 @@
 
 import tornado.gen
 from tornado.web import RequestHandler
+import tornado.escape
```
Comment (bnaul) [inline cesium_app/handlers/prediction.py]: Maybe just import `json_decode` directly so those calls will be shorter?

File: cesium_app/handlers/prediction.py [hunk 2]
```diff
@@ -137,3 +139,28 @@ def delete(self, prediction_id):
         prediction = self._get_prediction(prediction_id)
         prediction.delete_instance()
         return self.success(action='cesium/FETCH_PREDICTIONS')
+
+
+class PredictSingleHandler(BaseHandler):
+    def post(self):
+        ts_data = tornado.escape.json_decode(self.get_argument('ts_data'))
+        model_id = int(self.get_argument('modelID'))
+        features_to_use = tornado.escape.json_decode(
```
Comment (bnaul) [inline cesium_app/handlers/prediction.py]: Aren't the features to use stored in the model? The user shouldn't be able to choose this.

File: cesium_app/handlers/prediction.py [hunk 3]
```diff
@@ -137,3 +139,28 @@ def delete(self, prediction_id):
...
+
+        if features_to_use == 'all':
+            features_to_use = CADENCE_FEATS + GENERAL_FEATS + LOMB_SCARGLE_FEATS
+
+        model_uri = Model.get(Model.id == model_id).file.uri
+        model = joblib.load(model_uri)
+
+        fset_data = cesium.featurize.featurize_time_series(
+            *ts_data, features_to_use=features_to_use, meta_features=meta_feats)
+        fset = cesium.featureset.Featureset(fset_data).impute()
```
Comment (bnaul) [inline cesium_app/handlers/prediction.py]: Should imputation strategy be an option that the user can POST? I can't even remember what the default is actually...

**Comment by stefanv** [comment]
I just chatted with Josh about this, and he says it would be nice if we have the ability to persist the result (i.e., automatically go through the steps of "data -> featurize -> model -> predict".

**Comment by stefanv** [comment]
Also, I guess we want to use the features for the model we're predicting on, not all of them?

**Comment by acrellin** [comment]
@stefanv can you clarify what you mean by automatically go through the steps?

**Comment by acrellin** [comment]
@stefanv it's not using all of them, only if "all" is specified (as in the test)

**Comment by stefanv** [comment]
Just persist the resulting dataset, featureset, model, and prediction.

**Comment by acrellin** [comment]
@stefanv so scrapping this approach altogether?

**Comment by acrellin** [comment]
@bnaul I wanted to leave room for the case where a model may exist, but doesn't have an associated `featureset` (which the approach you suggested assumes it has), e.g. if one creates the DB state manually with just a pre-built model

**Comment by acrellin** [comment]
@bnaul implemented your other two suggestions, though. thanks for the feedback

**Comment by bnaul** [comment]
Hm...that still makes me a bit uneasy because you could get into a situation where a model was trained on `['amplitude', 'minimum']` but then for new data we're feeding it `['amplitude', 'maximum']` and no error is raised.

**Comment by acrellin** [comment]
@bnaul Yeah, I see your point. The two routes I see are 1) requiring that the `model` have an associated `featureset`, or 2) adding an optional `features_list` attribute to the `Model` model, so this will still work in the case I described above.

**Comment by bnaul** [comment]
The case you described is impossible right now, though; were we going to add support for that for some reason?

```
Foreign-key constraints:
    "model_featureset_id_fkey" FOREIGN KEY (featureset_id) REFERENCES featureset(id) ON DELETE CASCADE
```

**Comment by acrellin** [comment]
I had been thinking about it, but think it's fine to go with option 1) for now. I'll update this PR to accord with that.

**Review by bnaul — APPROVED** [review]

File: cesium_app/app_server.py [hunk 4]
```diff
@@ -32,6 +33,7 @@ def make_app():
         (r'/models(/.*)?', ModelHandler),
         (r'/predictions(/[0-9]+)?', PredictionHandler),
         (r'/predictions/([0-9]+)/(download)', PredictionHandler),
+        (r'/predict_raw_data(/.*)?', PredictRawDataHandler),
```
Comment (stefanv) [inline cesium_app/app_server.py]: remove the `(/.*)?`?

File: cesium_app/tests/frontend/test_predict_raw_data.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import requests
+import json
+
+
+def test_predict_raw_data():
+    with create_test_project() as p, create_test_dataset(p) as ds,\
+         create_test_featureset(p) as fs, create_test_model(fs) as m:
+        ts_data = [[1, 2, 3, 4], [32.2, 53.3, 32.3, 32.52], [0.2, 0.3, 0.6, 0.3]]
+        impute_kwargs = {'strategy': 'constant', 'value': None}
+        query_string = ('http://localhost:5000/predict_raw_data?ts_data={}'
```
Comment (stefanv) [inline cesium_app/tests/frontend/test_predict_raw_data.py]: Let's not hardcode things like ports---I use a different port to test on my machine.

Comment (stefanv) [reply cesium_app/tests/frontend/test_predict_raw_data.py]: Also, let's not do a query but handle this as a POST.

**Pull request merged**